### PR TITLE
Fix pointer mismatchings and other warnings

### PR DESCRIPTION
**Pull request by AtariDreams**
Many of these castings are unneeded if we change the variables to work better with each other.

Note that none of the public facing functions had their signature changed.

### CODE REVIEW

File: lib/base64.c [hunk 1]
```diff
@@ -55,10 +55,10 @@ static const char base64url[]=
 static size_t decodeQuantum(unsigned char *dest, const char *src)
 {
   size_t padding = 0;
-  const char *s;
+  const char *s = src;
   unsigned long i, x = 0;
 
-  for(i = 0, s = src; i < 4; i++, s++) {
+  for(i = 0; i < 4; i++, s++) {
```
Comment (bagder) [inline lib/base64.c:61]: This is not a pointer mismatch change.

File: lib/connect.c [hunk 2]
```diff
@@ -409,7 +409,7 @@ static CURLcode bindlocal(struct Curl_easy *data,
                Curl_printable_address. The latter returns only numeric scope
                IDs and the former returns none at all.  So the scope ID, if
                present, is known to be numeric */
-            si6->sin6_scope_id = atoi(scope_ptr);
+            si6->sin6_scope_id = (unsigned int)strtoul(scope_ptr, NULL, 10);
```
Comment (bagder) [inline lib/connect.c:412]: if we change this, shouldn't we also make a proper overflow check as well? After all, `strtoul()` might return a value that cannot be stored as a scope id...

File: lib/connect.c [hunk 3]
```diff
@@ -866,7 +866,7 @@ CURLcode Curl_is_connected(struct Curl_easy *data,
   int error = 0;
   struct curltime now;
   int rc = 0;
-  unsigned int i;
+  int i;
```
Comment (bagder) [inline lib/connect.c:869]: why?

File: lib/connect.c [hunk 4]
```diff
@@ -1056,8 +1056,8 @@ CURLcode Curl_is_connected(struct Curl_easy *data,
       hostname = conn->http_proxy.host.name;
     else
 #endif
-      if(conn->bits.conn_to_host)
-        hostname = conn->conn_to_host.name;
+    if(conn->bits.conn_to_host)
+      hostname = conn->conn_to_host.name;
```
Comment (bagder) [inline lib/connect.c:1060]: why?

File: lib/curl_fnmatch.c [hunk 5]
```diff
@@ -71,18 +71,19 @@ typedef enum {
...
   parsekey_state state = CURLFNM_PKW_INIT;
 #define KEYLEN 10
   char keyword[KEYLEN] = { 0 };
-  int found = FALSE;
-  int i;
-  unsigned char *p = *pattern;
+  bool found = FALSE;
+  unsigned int i;
+  char *p = *pattern;
+  char c;
```
Comment (bagder) [inline lib/curl_fnmatch.c:82]: Why the changed scope for `c` here?

File: lib/curl_fnmatch.c [hunk 6]
```diff
@@ -140,33 +141,33 @@ static char_class charclass(unsigned char c)
...
-    unsigned char endrange = *p++;
+    char endrange = *p++;
 
     if(endrange == '\\')
       endrange = *p++;
     if(endrange >= c && charclass(endrange) == cc) {
       while(c++ != endrange)
         if(charclass(c) == cc)  /* Chars in class may be not consecutive. */
-          charset[c] = 1;
+          charset[(unsigned int)c] = 1;
```
Comment (bagder) [inline lib/curl_fnmatch.c:159]: This change **introduces** two typecasts. Why is that better?

File: lib/curl_fnmatch.c [hunk 7]
```diff
@@ -245,7 +246,7 @@ static int setcharset(unsigned char **p, unsigned char *charset)
       if(c == ']')
         return SETCHARSET_OK;
       state  = CURLFNM_SCHS_DEFAULT;
-      charset[c] = 1;
+      charset[(unsigned int)c] = 1;
```
Comment (bagder) [inline lib/curl_fnmatch.c:249]: Here too, you add **seven** typecasts. Why is this better?

**Review by bagder — COMMENTED** [review]
I review a part of this PR and it made me question what you are doing so I paused.

File: lib/curl_fnmatch.c (hunk 6 above)
Comment (AtariDreams) [reply lib/curl_fnmatch.c:159]: In case c is negative

File: lib/curl_fnmatch.c (hunk 7 above)
Comment (AtariDreams) [reply lib/curl_fnmatch.c:249]: In case c is negative

File: lib/connect.c (hunk 3 above)
Comment (AtariDreams) [reply lib/connect.c:872]: const int other = i ^ 1;

File: lib/curl_fnmatch.c (hunk 5 above)
Comment (AtariDreams) [reply lib/curl_fnmatch.c:82]: c is defined elsewhere, causing shadowing

File: lib/connect.c (hunk 4 above)
Comment (AtariDreams) [reply lib/connect.c]: Because the formatting here is broken

File: lib/connect.c (hunk 3 above)
Comment (bagder) [reply lib/connect.c:874]: and why does that get better if `i` is not unsigned?

File: lib/connect.c (hunk 4 above)
Comment (bagder) [reply lib/connect.c:1065]: in your opinion then perhaps, but not in mine. The `if` is intended based on the `else` on the line before.

File: lib/connect.c (hunk 3 above)
Comment (AtariDreams) [reply lib/connect.c:874]: Because I only goes up to 1 anyway. It silences warnings, assuming you intended to use the variable like a Boolean.

Comment (bagder) [reply lib/connect.c:874]: What warning? There's nothing wrong there, having a warning for that seems overzealous.

File: lib/curl_fnmatch.c (hunk 6 above)
Comment (bagder) [inline lib/curl_fnmatch.c:159]: I still don't see the point in this change. It still *adds* two typecasts, that's contrary to the main motivation for this PR.

Comment (AtariDreams) [reply lib/curl_fnmatch.c:159]: It has to be because char is signed

**Comment by bagder** [comment]
[quite bad](https://github.com/curl/curl/actions/runs/3352731574/jobs/5555067542)

**Comment by AtariDreams** [comment]
Not anymore

File: tests/server/util.c [hunk 8]
```diff
@@ -378,18 +378,16 @@ static char raw_toupper(char in)
...
     }
     max--;
     first++;
     second++;
   }
-  if(0 == max)
-    return 1; /* they are equal this far */
 
-  return raw_toupper(*first) == raw_toupper(*second);
+  return 1; /* they are equal this far */
```
Comment (bagder) [inline tests/server/util.c:390]: this change is completely unrelated to the topic and purpose of this PR

**Comment by bagder** [comment]
This PR starts to smell like a million tiny changes you think should be done, but most of them are now not about "pointer mismatches and other warnings". It's a busload of various unrelated changes that should not be in the same PR. It makes this PR very hard to review and next to impossible to ultimately merge.

**Comment by jay** [comment]
You must have very strict compiler settings for the warnings you are seeing. Many of these changes are not needed.

We allow counter variables to go negative or wrap around. For example this is acceptable:

int foo = 5;
while(foo--) {
}
return

We allow subscript values to be a signed integer type. For example this is acceptable:

int foo = 5;
bar[foo]

We allow char * function parameters.

I'm closing this PR in favor of #9835 which is what is left after my review.
